telegram/passport/credentials.py

```diff
@@ -0,0 +1,394 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+import binascii
+import json
+from base64 import b64decode
+
+from future.utils import bord
+
+try:
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives.asymmetric.padding import OAEP, MGF1
+    from cryptography.hazmat.primitives.ciphers import Cipher
+    from cryptography.hazmat.primitives.ciphers.algorithms import AES
+    from cryptography.hazmat.primitives.ciphers.modes import CBC
+    from cryptography.hazmat.primitives.hashes import SHA512, SHA256, Hash, SHA1
+
+    CRYPTO = True
+except ImportError:
+    CRYPTO = False
+
+from telegram import TelegramObject, TelegramError
+
+
+class TelegramDecryptionError(TelegramError):
+    pass
+
+
+def decrypt(secret, hash, data):
+    """
+    Decrypt per telegram docs at https://core.telegram.org/passport.
+
+    Args:
+        secret (:obj:`str` or :obj:`bytes`): The encryption secret, either as bytes or as a
+            base64 encoded string.
+        hash (:obj:`str` or :obj:`bytes`): The hash, either as bytes or as a
+            base64 encoded string.
+        data (:obj:`str` or :obj:`bytes`): The data to decrypt, either as bytes or as a
+            base64 encoded string.
+
+    Raises:
+        :class:`TelegramDecryptionError`: Raised if the given hash does not match the hash of
+            decrypted data
+
+    Returns:
+        :obj:`bytes`: The decrypted data as bytes
+
+    """
+    try:
+        secret = b64decode(secret)
+    except (binascii.Error, TypeError):
+        pass
+    try:
+        hash = b64decode(hash)
+    except (binascii.Error, TypeError):
+        pass
+    try:
+        data = b64decode(data)
+    except (binascii.Error, TypeError):
+        pass
+    digest = Hash(SHA512(), backend=default_backend())
+    digest.update(secret + hash)
+    secret_hash_hash = digest.finalize()
+    key, iv = secret_hash_hash[:32], secret_hash_hash[32:32 + 16]
+    cipher = Cipher(AES(key), CBC(iv), backend=default_backend())
+    decryptor = cipher.decryptor()
+    data = decryptor.update(data) + decryptor.finalize()
+    digest = Hash(SHA256(), backend=default_backend())
+    digest.update(data)
+    data_hash = digest.finalize()
+    if data_hash != hash:
+        raise TelegramDecryptionError("Hashes are not equal! {} != {}".format(data_hash, hash))
```

Comment:
The bot completely hangs up when this error occurs, because it is raised when receiving the update. Even a bot restart does not fix it.

Reply:
I'm not a 100% sure that the test I wrote tests this completely... could you try your test with my latest commits?